keps/sig-instrumentation/20200415-cardinality-enforcement.md

```diff
@@ -0,0 +1,276 @@
+
+### Non-Goals
+
+We will expose the machinery and tools to bind a metric's labels to a discrete set of values.
+
+It is *not a goal* to implement and plumb this solution for each Kubernetes component (there are many SIGs and a number of verticals, which may have their own preferred way of doing things). As such it will be up to component owners to leverage this functionality that we provide, by feeding configuration data through whatever mechanism deemed appropriate (i.e. command line flags or reading from a file).
+
+## Proposal
+
+TLDR; we want to be able to *dynamically configure* a whitelist of label values for a metric.
+
+By *dynamically configure*, we mean configure a whitelist *at runtime* rather than during build/compile step (this is so ugh).
+
+And by *at runtime*, we mean, more specifically, during the boot sequence for a Kubernetes component (and we mean daemons here, not cli tools like kubectl).
+
+Brief aside: a Kubernetes component (which is a daemon) is an executable, which you can launch from the command line manually if you so desired. Components take a number of start-up configuration flags, which are passed into the component to modify execution paths (if curious, you can check out the [zillion flags we have on the kubelet](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/)). It is also possible to read configuration data from files (like yamls) during the component boot sequence. This KEP does not have an opinion on the specific mechanism used to load config data into a Kubernetes binary during the boot sequence. What we *actually* care about, is just the fact that it is possible.
+
+Our design, is thus config-ingestion agnostic.
+
+Our design is also based on the premise that metrics can be uniquely identified (i.e. by their metric descriptor). Fortunately for us, this is actually a built in constraint of prometheus clients (which is how we instrument the Kubernetes components). This metric ID is resolvable to a unique string (this is pretty evident when you look at a raw prometheus client endpoint).
+
+We want to provide a data structure which basically maps a unique metric ID and one of it's labels to a bounded set of values.
+
+Purely for illustration, take the following metric:
+
+```json
+[
+	{
+		"metric-id": "some_metric"
+	}
+]
+
+```
+
+We want to be able to express this metric and target a label:
+
+
+```json
+[
+	{
+		"metric-id": "some_metric",
+		"label": "label_too_many_values",
+	}
+]
+
+```
+
+And we want to express a set of expected values:
+
+
+```json
+[
+	{
+		"metric-id": "some_metric",
+		"label": "label_too_many_values",
+		"labelValueWhitelist": [
+			"1",
+			"2",
+			"3"
+		]
+	}
+]
+
+```
+
+Since we already have an interception layer built into our Kubernetes monitoring stack (from the metrics stability effort), we can leverage existing wrappers to provide a global entrypoint for intercepting and enforcing rules for individual metrics.
```

Comment:
Since we already have an interception layer, why not simply enforce the cardinality limit there directly?

If we could specify that a certain metric allows up to N unique label sets (or K values per label), an attempt to add anything beyond that would just use some predefined constant instead. This has the downside (compared to your proposal) of being less predictable, because you never know what will get filtered out. However, in practice if K/N are sufficiently high, this shouldn't be a concern (all frequently used label values will show up early, so will be reported). Additionally, the approach I'm suggesting will:
- enforce that your metrics "just work" instead of putting cluster admins in a position where they have to react based on alerts, when something is already broken.
- allow monitoring entities that are of high cardinality, but when interesting label values cannot be listed up front.

If the "not predictable enough" bit is problematic, we can also combine both ideas and treat whitelists as "always working" and enforce explicit cardinality limits only on values that were not whitelisted. This would slightly change the semantics of whitelisting - lack of a whitelist would be equivalent to an empty whitelist.

Reply:
We considered this for the recent security vulnerability. The problem is exactly as you describe which is it is 'not predictable enough'. We basically lose deterministic metric output if we adopt this (not to mention metric fidelity). While the remaining metrics will 'work' in the sense that they will be bound to N number of dimensions and will thus not cause memory leaks, they will not be reliable, since the N number of dimensions you end up getting will be determined at runtime. 

The alternative you mention is problematic from a practical perspective (it is an alternative way of implementing the thing that @dashpole mentioned in [an earlier comment](https://github.com/kubernetes/enhancements/pull/1692#discussion_r409877573)). We'd effectively be blanking out all label values for every metric in the Kubernetes codebase until people explicitly laid out a whitelist. That means every single metric would have to be audited, which is quite invasive. 

Alternatively, we could use a ratcheting approach, where we enforce that all **_new_** metrics must have a whitelist explicitly specified (this could potentially be a requirement of a metric being promoted to the STABLE class).

Reply:
You can think of my proposal as an extension to yours, such that instead of getting `label_too_many_values` right away, this would still work until certain label cardinality limit is reached. Whitelisting would guarantee a value will not be dropped, but other values wouldn't be dropped either unless there is too many of them. Cluster admin can configure the per metric and per label limits once and can get alerted on "some metric labels are dropped" instead of "your metrics storage is getting blown up".

Reply:
I think the idea of specifying a cardinality limit instead of explicit whitelist has a lot of benefits.  It allows us to potentially turn any memory leak, which is a DOS security vulnerability, into a degradation of metrics, which is easy to monitor.

What if we adopted a form of both approaches?  We could have a global per-label cardinality limit e.g. 1000 values.  IMO it shouldn't even be configurable.  It should be high enough that we are confident such a label is a "bug".  This would bound the number of metric streams at compile time since metrics, labels, and now label values are all bounded.  However, we should have some way for operators to "save" their metrics when labels start being messed up by the cardinality limit.  This is where this proposal could help.  It doesn't allow exceeding the cardinality limit; it just allows specifying which labels are kept.

Reply:
Having a cardinality limit plus a whitelist is essentially what I'm suggesting. I think we should have both per-label and per-metric though. 1000 label values on every label works fine when there is a single problematic label, but without per-metric limits, any metric having multiple labels can still eat a ton of memory.

This would be slightly more complicated to implement correctly than purely per-label limiting approach, but would give us a realistic hard memory limit for every metric. If I have 10 labels, each of them allowing up to 1000 values, the number of possible label combinations is bound, but we can eat entire memory on any machine anyway.

Reply:
@x13n's comment is actually why I don't want to have the limit. Something like 1000 series limit isn't really going to help since it's the multiplicative effect of the label values which cause cardinality issues. 

Also, I am personally wary of setting a limit for something like request_total or request_duration since they are disproportionately large metrics relevant to the norm. We have like 50 buckets on durations, so I suspect we hit 1k timeseries pretty easily/often.

Reply:
Regardless, it's a slightly orthogonal dimension to the KEP. It sounds like people are actually on board with the label whitelist approach, doing so will not prevent us from also doing stuff like introducing label limits in the future (it's not something I am completely convinced of at the moment).

Reply:
yep.  I was thinking that as well.  We can definitely consider each (label whitelist or cardinality limit) separately, and i'm in favor of the current proposed whitelist mechanism.

Reply:
SGTM, limiting the cardinality can be done (or not) in a separate proposal, to switch from reacting to preventing issues. Whitelisting itself is indeed already giving cluster admins a way of manually mitigating large metric issues on selected metrics.